torch/lib/THCUNN/LookupTable.cu

```diff
@@ -162,36 +162,76 @@ __global__ void cunn_LookupTable_accGradParametersKernel(
   }
 }
 
-/*
- * Keep the norm of weight smaller than maxNorm
- */
 template <typename Dtype, typename Acctype>
-struct pow_v
-{
-  Acctype normType;
-  pow_v(Dtype v) : normType(ScalarConvert<Dtype, Acctype>::to(v)) {}
-  __host__ __device__
-  Acctype operator()(const Dtype& x) const {
+__device__
+Acctype fast_pow(Dtype x, Acctype normType) {
     Acctype xA = ScalarConvert<Dtype, Acctype>::to(x);
     if (normType == 1)
       return std::abs(xA);
     else if (normType == 2)
       return xA * xA;
     else
       return std::pow(std::abs(xA), normType);
-  }
-};
+}
+
+template <typename Dtype>
+__device__
+Dtype get_padded_value(Dtype *weight_ptr, THCIndex_t idx, int64_t dim) 
+{
+  if (idx < dim)
+    return weight_ptr[idx];
+  else
+    return 0;
+}
 
-template <typename T>
-struct multiply_s
+/* Calculate norms of the rows of weight_ptr given by idx_ptr and capture them in norms */
+template <typename Dtype, typename Acctype>
```

Comment:
template this kernel based on IndexType which is either `unsigned int` or `unsigned long`, as 64-bit integer math on the GPU is really slow, and at launch time figure out the maximum index that the kernel would need, and launch one version or the other based on this (you can find similar things in the cutorch code, there are utility functions for this).

Then, use IndexType in the kernel instead of int64_t.